File: frontend/src/api/manufacturers.ts
/**
 * Manufacturers API
 * =================
 * Handles all manufacturer-related API calls.
 */

import { 
  Manufacturer, 
  ManufacturerStats, 
  ManufacturerMedicine, 
  ManufacturerInventory, 
  ApiResponse 
} from '@/types';

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:5000';

class ManufacturersAPI {
  private baseUrl: string;

  constructor() {
    this.baseUrl = `${API_BASE_URL}/api/manufacturers`;
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    try {
      const response = await fetch(`${this.baseUrl}${endpoint}`, {
        headers: {
          'Content-Type': 'application/json',
          ...options.headers,
        },
        ...options,
      });

      const data = await response.json();

      if (!response.ok) {
        return {
          success: false,
          error: data.error || `HTTP ${response.status}`,
        };
      }

      return data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Network error',
      };
    }
  }

  /**
   * Get all manufacturers with pagination
   */
  async getAllManufacturers(
    page: number = 1,
    limit: number = 50,
    activeOnly: boolean = true
  ): Promise<ApiResponse<{ data: Manufacturer[]; pagination: any }>> {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
      active_only: activeOnly.toString(),
    });

    return this.request<{ data: Manufacturer[]; pagination: any }>(`/?${params}`);
  }

  /**
   * Search manufacturers by query
   */
  async searchManufacturers(
    query: string,
    activeOnly: boolean = true
  ): Promise<ApiResponse<Manufacturer[]>> {
    const params = new URLSearchParams({
      query,
      active_only: activeOnly.toString(),
    });

    return this.request<Manufacturer[]>(`/search?${params}`);
  }

  /**
   * Get manufacturer by ID
   */
  async getManufacturerById(manufacturerId: number): Promise<ApiResponse<Manufacturer>> {
    return this.request<Manufacturer>(`/${manufacturerId}`);
  }

  /**
   * Create a new manufacturer
   */
  async createManufacturer(data: Partial<Manufacturer>): Promise<ApiResponse<Manufacturer>> {
    return this.request<Manufacturer>('/', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  /**
   * Update manufacturer information
   */
  async updateManufacturer(
    manufacturerId: number,
    data: Partial<Manufacturer>
  ): Promise<ApiResponse<Manufacturer>> {
    return this.request<Manufacturer>(`/${manufacturerId}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  }

  /**
   * Delete manufacturer (soft delete)
   */
  async deleteManufacturer(manufacturerId: number): Promise<ApiResponse<{ message: string }>> {
    return this.request<{ message: string }>(`/${manufacturerId}`, {
      method: 'DELETE',
    });
  }

  /**
   * Get manufacturer statistics
   */
  async getManufacturerStatistics(manufacturerId: number): Promise<ApiResponse<ManufacturerStats>> {
    return this.request<ManufacturerStats>(`/${manufacturerId}/statistics`);
  }

  /**
   * Get medicines manufactured by this manufacturer
   */
  async getManufacturerMedicines(manufacturerId: number): Promise<ApiResponse<ManufacturerMedicine[]>> {
    return this.request<ManufacturerMedicine[]>(`/${manufacturerId}/medicines`);
  }

  /**
   * Get inventory levels for manufacturer's medicines
   */
  async getManufacturerInventory(manufacturerId: number): Promise<ApiResponse<ManufacturerInventory[]>> {
    return this.request<ManufacturerInventory[]>(`/${manufacturerId}/inventory`);
  }
}

// Export singleton instance
export const manufacturersAPI = new ManufacturersAPI();

File: backend/routes/manufacturers.py
"""
Manufacturers API Routes
========================
Handles all manufacturer-related API endpoints.
"""

from flask import Blueprint, request, jsonify
import logging
from services.manufacturer_service import manufacturer_service

logger = logging.getLogger(__name__)

# Create blueprint
manufacturers_bp = Blueprint('manufacturers', __name__, url_prefix='/api/manufacturers')


@manufacturers_bp.route('/', methods=['GET'])
def get_all_manufacturers():
    """Get all manufacturers with pagination."""
    try:
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 50))
        active_only = request.args.get('active_only', 'true').lower() == 'true'
        
        result = manufacturer_service.get_all_manufacturers(page, limit, active_only)
        return jsonify(result)
        
    except Exception as e:
        logger.error(f"Error in get_all_manufacturers: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/search', methods=['GET'])
def search_manufacturers():
    """Search manufacturers by query."""
    try:
        query = request.args.get('query', '')
        active_only = request.args.get('active_only', 'true').lower() == 'true'
        
        if not query:
            return jsonify({
                'success': False,
                'error': 'Query parameter is required'
            }), 400
        
        result = manufacturer_service.search_manufacturers(query, active_only)
        return jsonify(result)
        
    except Exception as e:
        logger.error(f"Error in search_manufacturers: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/<int:manufacturer_id>', methods=['GET'])
def get_manufacturer_by_id(manufacturer_id):
    """Get manufacturer by ID."""
    try:
        result = manufacturer_service.get_manufacturer_by_id(manufacturer_id)
        
        if result['success']:
            return jsonify(result)
        else:
            return jsonify(result), 404
            
    except Exception as e:
        logger.error(f"Error in get_manufacturer_by_id: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/', methods=['POST'])
def create_manufacturer():
    """Create a new manufacturer."""
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({
                'success': False,
                'error': 'Request body is required'
            }), 400
        
        result = manufacturer_service.create_manufacturer(data)
        
        if result['success']:
            return jsonify(result), 201
        else:
            return jsonify(result), 400
            
    except Exception as e:
        logger.error(f"Error in create_manufacturer: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/<int:manufacturer_id>', methods=['PUT'])
def update_manufacturer(manufacturer_id):
    """Update manufacturer information."""
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({
                'success': False,
                'error': 'Request body is required'
            }), 400
        
        result = manufacturer_service.update_manufacturer(manufacturer_id, data)
        
        if result['success']:
            return jsonify(result)
        else:
            return jsonify(result), 400
            
    except Exception as e:
        logger.error(f"Error in update_manufacturer: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/<int:manufacturer_id>', methods=['DELETE'])
def delete_manufacturer(manufacturer_id):
    """Deactivate manufacturer (soft delete)."""
    try:
        result = manufacturer_service.deactivate_manufacturer(manufacturer_id)
        
        if result['success']:
            return jsonify(result)
        else:
            return jsonify(result), 404
            
    except Exception as e:
        logger.error(f"Error in delete_manufacturer: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/<int:manufacturer_id>/statistics', methods=['GET'])
def get_manufacturer_statistics(manufacturer_id):
    """Get manufacturer statistics."""
    try:
        result = manufacturer_service.get_manufacturer_statistics(manufacturer_id)
        
        if result['success']:
            return jsonify(result)
        else:
            return jsonify(result), 404
            
    except Exception as e:
        logger.error(f"Error in get_manufacturer_statistics: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/<int:manufacturer_id>/medicines', methods=['GET'])
def get_manufacturer_medicines(manufacturer_id):
    """Get medicines manufactured by this manufacturer."""
    try:
        result = manufacturer_service.get_manufacturer_medicines(manufacturer_id)
        
        if result['success']:
            return jsonify(result)
        else:
            return jsonify(result), 404
            
    except Exception as e:
        logger.error(f"Error in get_manufacturer_medicines: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@manufacturers_bp.route('/<int:manufacturer_id>/inventory', methods=['GET'])
def get_manufacturer_inventory(manufacturer_id):
    """Get inventory levels for manufacturer's medicines."""
    try:
        result = manufacturer_service.get_manufacturer_inventory(manufacturer_id)
        
        if result['success']:
            return jsonify(result)
        else:
            return jsonify(result), 404
            
    except Exception as e:
        logger.error(f"Error in get_manufacturer_inventory: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500